File: Strings/java/InvalidTransactions.java
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.ArrayList;
import java.util.HashSet;

public class InvalidTransactions {

    public static List<String> invalidTransactions(String[] transactions) {

        List<String> invalidList = new ArrayList<>();
        int size = transactions.length;
        boolean[] booleanArr = new boolean[size];

        
        for(int i=0; i<size; i++) {
            String[] data = transactions[i].split(",");

            if(Double.parseDouble(data[2]) > 1000 ) {
                booleanArr[i] = true;
            }

            for(int j=i+1; j<size; j++) {
                String[] data2 = transactions[j].split(",");

                if(data[0].equals(data2[0]) && !data[3].equals(data2[3])) {
                    if(Math.abs(Double.parseDouble(data[1]) - Double.parseDouble(data2[1])) <= 60) {
                        booleanArr[j] = true;
                        booleanArr[i] = true;
                    }
                }
            }
        }

        for(int i=0; i<size; i++ ) {
            if(booleanArr[i]) {
                invalidList.add(transactions[i]);
            }
        }
        return invalidList;
    }

    public static void main(String[] args) {
        String[] arr = {"alice,20,800,mtv","bob,50,1200,mtv"};
        String[] arr2 = { "alice,20,800,mtv", "alice,50,100,beijing" };
        String[] arr3 = { "alice,20,800,mtv", "alice,50,100,mtv", "alice,51,100,frankfurt" };
        String[] arr4 = { "alice,20,800,mtv","alice,50,1200,mtv"};
        System.out.println(invalidTransactions(arr).toString());
        System.out.println(invalidTransactions(arr2).toString());
        System.out.println(invalidTransactions(arr3).toString());
        System.out.println(invalidTransactions(arr4).toString());
    }

}

// ["alice,20,800,mtv","alice,50,100,mtv","alice,51,100,frankfurt"]

File: Strings/javascript/invalidTransactions.js
/**
 * 
 * @param {Array<String>} transactions
 *  Lista de transações, sendo que cada posição é uma String contendo
 *  {name}, {time}, {value}, {city}, separados por vírgulas.
 *  
 *  Critérios para invalidar: value > 1000 OU mesma pessoa realizar uma transação
 *  em 60 ou menos minutos em duas cidades diferentes.
 * 
 *  @return {Array<String>} lista de todas as transações inválidas. 
 */

class Transaction {

    constructor(csv) {
        const data = csv.split(",");
        this.name = data[0];
        this.time = parseInt(data[1]);
        this.value = parseInt(data[2]);
        this.city = data[3];
    }
}
function invalidTransactions(transactions) {

    //Inicia uma lista de booleanos com valores falsos
    let invalidPositions = [];
    const size = transactions.length;
    for(let i=0; i<size; i++) {
        invalidPositions[i] = false;
    }


    for(let i=0; i<size; i++) {

        const t = new Transaction(transactions[i]);

        if(t.value > 1000) invalidPositions[i] = true;

        for(let j=i+1; j<size; j++) {

            const u = new Transaction(transactions[j]);

            if(t.name == u.name && t.city != u.city && Math.abs(t.time - u.time) <= 60) {
                invalidPositions[i] = true;
                invalidPositions[j] = true;
            }
        }
        
    }
    let invalidList = [];
    for(let i=0; i<size; i++) {
        if(invalidPositions[i]) invalidList.push(transactions[i]);
    }

    return invalidList;
}

console.log(invalidTransactions(["alice,20,800,mtv", "alice,50,100,mtv", "alice,51,100,frankfurt"]));